pdl_syntax.md

```diff
@@ -0,0 +1,828 @@
+ */
+fixed MD5 16
+```
+
+com/linkedin/pegasus/generator/examples/StringList.pdl
+
+```
+namespace com.linkedin.pegasus.generator.examples
+
+/**
+ * A list of strings
+ */
+record StringList {
+  element: string
+  next: optional 
+
+    /**
+     * A list of strings
+     */
+    record StringList {
+      element: string
+      next: optional StringList
+    }
+}
+```
+
+com/linkedin/pegasus/generator/examples/InlinedExample.pdl
+
+```
+namespace com.linkedin.pegasus.generator.examples
+
+/**
+ * Example on how you can declare an enum and a record inside another record
+ */
+record InlinedExample {
+
+  /**
+   * This is how we inline enum declaration without creating a new pdl file
+   */
+  @symbolDocs.FOO = "It's a foo!"
+  @symbolDocs.NONE = "It's a baz!"
+  @symbolDocs.HASH = "It's a bar!"
+  myEnumField: enum EnumDeclarationInTheSameFile {
+    FOO
+    BAR
+    BAZ
+  }
+
+  /**
+   * A regular string
+   */
+  stringField: string
+
+  /**
+   * A regular int
+   */
+  intField: int
+
+  /**
+   * In this example we will declare a record and an enum inside a union
+   */
+  UnionFieldWithInlineRecordAndEnum: optional union[record myRecord {
+
+    /**
+     * random int field
+     */
+    foo1: int
+
+    /**
+     * random int field
+     */
+    foo2: int
+  }, 
+    /**
+     * Random enum
+     */
+    enum anotherEnum {
+
+      /**
+       * description about FOOFOO
+       */
+      FOOFOO
+
+      /**
+       * description about BARBAR
+       */
+      BARBAR
+    }]
+}
+```
+
+## Record Type
+Records contain any number of fields, which can be primitive type, enums, unions, maps and arrays.
+
+A basic record type can contain a few fields.
+
+For example:
+```
+import org.example.time.DateTime
+
+record Example {
+  field1: string
+  field2: optional int
+  field3: DateTime
+}
+```
+Record fields can be optional.
+
+For example :
+```
+namespace com.linkedin.pegasus.generator.examples
+
+/**
+ * A foo record
+ */
+record Foo {
+   field1: string
+   field2: optional string
+}
+```
+
+Record fields may have default values.
+
+For example :
+```
+namespace com.linkedin.pegasus.generator.examples
+
+/**
+ * A foo record
+ */
+record Foo {
+   field1: string
+   field2: string = "message"
+}
+```
+
+A optional field may have default value.
+
+For example : 
+```
+namespace com.linkedin.pegasus.generator.test.idl.records
+
+record WithOptionalPrimitiveDefault {
+  intWithDefault: optional int = 1
+}
+```
+
+### Doc Strings
+
+Types and fields may be documented using “doc strings”.
+
+For example : 
+```
+/**
+ * Doc strings may be added to types. This doc should describe the purposes
+ * of the Example type.
+ */
+record Example {
+  /**
+   * Doc strings may also be added to fields.
+   */
+   field1: string
+
+   /** Doc strings can be single line.*/
+   field2: int
+}
+```
+
+**Note:**
+If you use Java comment style for doc string, e.g "// Doc String", those doc strings will not be stored in in-memory schema.
+
+
+### Import
```

Comment:
Import should be a separate section outside records.

Reply:
Moved it to a separate section.